### ISSUE TITLE
Broken upgrades using edeliver and Ecto.Repo

### ISSUE THREAD

**Issue by ghedamat** [boldpoker/edeliver#68]
Hi, 

I've been struggling with what I think is a bug in edeliver for a while and I finally got to a small-ish reproduction case

TL;DR;

when `MyApp.Repo` is not supervised directly by the phoenix `Application` or when I have another OTP application that uses `MyApp.Repo`  if `edeliver` is used in the project the release `upgrade` process fails.

I've created a reproduction application (you'll need to setup postgres to make it work) with two branches and instructions on how to reproduce the bug

https://github.com/ghedamat/foo (see README for detailed description)

the only difference between `master` and `edeliver-test` is the fact that the latter adds the edeliver package and application.

I'm kinda stuck as I'm quite new to all this (otp/releases etc) so any help would be much appreciated

thanks a lot!

**Comment by bharendt** [comment]
Hi @ghedamat. I don't think this is a bug in edeliver, because the error occurs while the relup instructions are executed. This is pretty normal behaviour, because the relup instructions generated by the build tool (exrm in your case) are more like a default or template and often need to be adjusted. The other option is that wrong relup instrunctions are generated because of some misconfiguration (app file, supervision tree etc).
But maybe we can help to locate the error anyway. Thanks for providing a repository to reproduce that error, but for a quick investigation it would be nice if you could also provide the following four infos just here in the issue, which would be quite helpful (and are helpful in general to debug issuses during hot code upgrades):

 * The content of the relup file which is used to perform the upgrade (from `releases/0.0.2/relup`). 
 * The content of the app file which contains the names of the modules and registered names from `lib/foo-0.0.1/ebin/foo.app` and maybe also from `lib/foo-0.0.2/ebin/foo.app`
 * the supervision tree of your app, e.g. a screenshot from the [erlang observer app](http://erlang.org/doc/man/observer.html)
 * It looks like in addition to the error during the upgrade [some additional infos are logged](https://github.com/blackberry/Erlang-OTP/blob/master/lib/sasl/src/release_handler_1.erl#L567). This error message would help to identify the supervisor which cannot be suspended and the reason why.

**Comment by ghedamat** [comment]
@bharendt thanks for the quick response!

I'll try to provide this info as soon as I can. the only thing I had checked was if `relup` generated with and without `edeliver` being added to the project and they were identical.

maybe the difference lies in one of these other files (I'm slowly learning their purpose).

Will update you shortly. thanks again!

**Comment by ghedamat** [comment]
@bharendt here's the info you requested (I hope)

### without using edeliver (the version that works)
https://gist.github.com/ghedamat/757d11fc51a592f8f93a41d94cb7b0fb (relup and app files)

### with edeliver

https://gist.github.com/ghedamat/afcda26ee74f904b07032305e3ae25bd (relup, appfiles, and erlang.log)

### screenshot of my supervisor tree
https://gsnaps.s3.amazonaws.com/nonodenohost_2016-05-04_17-10-19.png
`Elixir.Foo.Endpoint` can't be seen but that's standard phoenix stuff

and related code https://github.com/ghedamat/foo/blob/master/lib/foo.ex#L12


### Notes

I was totally wrong and the relup files are not the same at all so likely the problem is there ?

I know it's a contrived example but it's to show that this happens (or at least I believe that's the case) only when I use `edeliver` and I do something with `Foo.Repo` that goes out of the default path.

Sorry if I'm explaining myself poorly and taking up your time, I'm quite convinced it's something I'm missing :/

thanks again!

(also happy to chat on irc/slack tomorrow or later in the week if that's easier)

**Comment by bharendt** [comment]
Ah, ok. Sorry. In your upgrade you are using the [automatic upgrade patching](https://github.com/boldpoker/edeliver/blob/master/lib/edeliver/relup/phoenix_modification.ex) feature of edeliver which means that edeliver could indeed be involved. 

Let's see... The error must occur somewhere after `==> Upgrading release to version 0.0.2...` and before `---> Upgrade to version 0.0.2 succeeded.`:

```erlang

{apply,
        {'Elixir.Edeliver.Relup.Instructions.Info',run,
            [<<"==> Upgrading release to version 0.0.2...">>]}},
    {load,{'Elixir.Foo.Endpoint',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.ErrorHelpers',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.ErrorView',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.Gettext',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.LayoutView',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.PageController',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.PageView',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.Repo',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.Router.Helpers',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.Router',soft_purge,brutal_purge}},
    {suspend,['Elixir.Foo.TestWorker']},
    {load,{'Elixir.Foo.TestWorker',soft_purge,brutal_purge}},
    {code_change,up,[{'Elixir.Foo.TestWorker',[]}]},
    {resume,['Elixir.Foo.TestWorker']},
    {load,{'Elixir.Foo.UserSocket',soft_purge,brutal_purge}},
    {load,{'Elixir.Foo.Web',soft_purge,brutal_purge}},
    {apply,
        {'Elixir.Edeliver.Relup.Instructions.Info',run,
            [<<"---> Upgrade to version 0.0.2 succeeded.">>]}},

```

and must have something to do with failing to suspend a process, and processes using the `Elixir.Foo.TestWorker` module are the onliest that are tried to be suspended there. And the error message `[error] release_handler: a which_children call to :ranch_acceptors_sup (#PID<0.1078.0>) was avoided. This supervisor is suspended and should likely be upgraded differently. Exiting ...` indicates that the release handler tries to suspend the `ranch_acceptors_sup` which is already suspended. Which is true, because edeliver modifies the relup in that way, that all [acceptors are suspended before loading the new code](https://github.com/boldpoker/edeliver/blob/master/lib/edeliver/relup/phoenix_modification.ex#L17) of the new release to avoid that new requests would crash because loading new modules is not atomic. This is also logged:

```
17:05:11.864 [info] ==> Preparing upgrade to version 0.0.2...
17:05:11.864 [info] Suspending ranch socket acceptors...
17:05:11.864 [info] Stopping 100 ranch socket acceptors...
17:05:11.866 [info] Suspended 100 ranch acceptors.
17:05:11.866 [info] Suspending ranch socket acceptor supervisor...
17:05:11.866 [info] Suspending phoenix websocket channels...

```

So the main question that remains is, why the release handler thinks it should suspend the `ranch_acceptors_sup` (which is even from a different application) when suspending actually the `Elixir.Foo.TestWorker`.

**Comment by bharendt** [comment]
Btw, if this issue prevents you from building upgrades for your production server, you have mainly the following three options:

  * edit the relup file using `edeliver edit relup` and remove the `{suspend,['Elixir.Foo.TestWorker']}`, `{code_change,up,[{'Elixir.Foo.TestWorker',[]}]}` and the `{resume,['Elixir.Foo.TestWorker']}` lines (or the corresponding instructions for your real upgrade)
  * pass your own relup-modification to the `--relup-mod=<module-name>` option of the `edeliver build upgrade` command, which could be derived from the [Edeliver.Relup.PhoenixModification](https://github.com/boldpoker/edeliver/blob/master/lib/edeliver/relup/phoenix_modification.ex) module and skip or change the part where the ranch acceptors are suspendend (e.g. [suspend the acceptors but not the supervisor](https://github.com/boldpoker/edeliver/blob/master/lib/edeliver/relup/instructions/suspend_ranch_acceptors.ex#L81))
  * or use the `--skip-relup-mod` opiton and skip the step where the relup file is auto-patched completely

But would be nice if we could find out the reason why the `ranch_acceptors_sup` is supsended a second time because of the 

```erlang
{suspend,['Elixir.Foo.TestWorker']}
```

relup instruction.

**Comment by ghedamat** [comment]
@bharendt thanks for this!

the workaround should help for now!


As an interesting point, I verified once again that moving `Foo.Repo` back up to the main application instead of using my custom supervisor generates a relup file that does NOT suspend `Foo.TestWorker`

see https://gist.github.com/6b0a171e9bcad6d5887d48ad21628245 
and https://github.com/ghedamat/foo/commit/68c1029657d766adcda691cc737e75bae5811e45


In my original project `Foo.Repo` is in the main application file but `Foo.TestWorker` uses an `Ecto.Model` and the same happens (relup tries to suspend the worker)

if nothing ecto-related is used the suspend is not present.


My hunch is that all this issue rotates around Ecto but that's where I stop for now :/

**Comment by ghedamat** [comment]
ok @bharendt sorry about the confusion

with the info you provided I simplified even more my reproduction case

this is the branch
https://github.com/ghedamat/foo/tree/edeliver-without-repo

this is the new application structure (I removed Ecto.Repo entirely out of the equation)
https://github.com/ghedamat/foo/blob/edeliver-without-repo/lib/foo.ex

and this is a gist with new relup, app files and error log
https://gist.github.com/ghedamat/1e0f196ff799107b49c0a5af81e17ed5

basically I "think" that if there's another supervisor involved this happens? let me know what else I could do to help hunt this down!

thanks again!

**Comment by ghedamat** [comment]
also, just to clarify

`mix edeliver build upgrade --with=0.0.2+11-e3981ef --skip-relup-mod` should generate a release with a relup file that is identical to the one that `exrm` would generate right?

it looks like the relup file is still modified.. 

editing it by hand works fine though, if I remove the suspend lines code is reloaded properly as you expected

**Comment by bharendt** [comment]
Yes, using `--skip-relup-mod` should not touch the relup at all and it should be the original one, which is generated by exrm. See [erlang-init:404](https://github.com/boldpoker/edeliver/blob/master/libexec/erlang-init#L404) and [modify_relup.ex:10](https://github.com/boldpoker/edeliver/blob/master/lib/exrm/plugins/modify_relup.ex#L10).

**Comment by bharendt** [comment]
But I think I can now explain why the upgrade fails: for a relup instruction like `{suspend,['Elixir.Foo.TestWorker']}` the release handler suspends all processes which use the `Elixir.Foo.TestWorker` as (a callback) module. According to the [documentation in the release handler](https://github.com/blackberry/Erlang-OTP/blob/master/lib/sasl/src/release_handler_1.erl#L482-L509) it does it by searching through the supervision tree, which is not possible if any supervisor is suspended. This implies, that no supervisor can be suspended before suspending / resuming any other process using the `suspend` relup instruction. To cope whith that restriction, we could add a relup-modification, that moves all `suspend` instructions before the suspending of any supervisor and all `resume` instructions after resuming any supervisor. I will add this to the default relup-modifications. So thank you very much for pointing this out!

**Comment by ghedamat** [comment]
thank you @bharendt ! 

I'll try again the `--skip-relup-mod` tomorrow and report back if I can't figure it out :p 

sorry for all the trouble and thanks again for the amazing help!

**Comment by bharendt** [comment]
@ghedamat, you were right. Unfortunately the `--skip-relup-mod` option was ignored and not forwarded to the mix task when building, [but this commit should fix that](https://github.com/boldpoker/edeliver/commit/23cdf71605d9f15b4716492c909862b8f4e30512):

**Comment by ghedamat** [comment]
@bharendt thanks for the quick fix!

I'll try that over the weekend for sure! and also thank you a lot for digging into a possible fix for the broader issue!

let me know if you wanna test the patch for that once you have something going.

I'll report back here in the meantime, thanks again!

**Comment by ghedamat** [comment]
@bharendt FYI the patch for `--skip-relup-mod` worked perfectly!

**Comment by bharendt** [comment]
Thanks! But we still need to fix the relup modification for phoenix apps. If it is not possible to suspend a supervisor (from the application supervision tree), the only way to avoid starting new children would be to suspend all processes which would start new children, even if this still does not prevent the supervisor from restarting children. But this does not make sense in case of a tcp connection process anyway. In the best case ranch would still accept new connection but wait (being suspended) to call any function of a callback module until the upgrade is finished.

**Comment by bharendt** [comment]
If I understand correctly, there are no ranch listeners nor ranch acceptors started dynamically, but just [a fixed number when ranch is started](https://github.com/ninenines/ranch/blob/ae8ff1fc0fcac71f24958962a362800775b43b41/src/ranch_acceptors_sup.erl#L48). I think in that case it is sufficient to just [remove the supsending of the supervisor](https://github.com/boldpoker/edeliver/commit/bc88f4d0b865e36898007fafd41109bb8bd13ef9):

**Comment by ghedamat** [comment]
@bharendt agreed. I was confirming that the simpler issue is indeed fixed but that leaves us with the bigger one.


I'm happy to try again later today with the commit you just pushed if you think it's ready for testing!

Thanks a bunch

**Comment by bharendt** [comment]
@ghedamat yes, would be nice if you could also confirm that the new changes work as expected. It is published as [hex package with version 1.2.5](https://hex.pm/packages/edeliver/1.2.5). Thanks.

**Comment by ghedamat** [comment]
@bharendt just did a deploy with version `1.2.5` and also checked the `relup` generated

worked like a charm!
